File: python3/brick_wall.py
"""
Space   : O(n)
Time    : O(wall layer * bricks)
"""


class Solution:
    def leastBricks(self, wall: List[List[int]]) -> int:
        if not wall:
            return 0

        d = {}

        for x in wall:
            width = 0
            for i, b in enumerate(x):
                width += b
                if i < len(x)-1:
                    if width not in d:
                        d[width] = 1
                    else:
                        d[width] += 1

        ans = len(wall)
        for _, v in d.items():
            ans = min(len(wall) - v, ans)

        return ans

File: golang/brick_wall.go
package main

func leastBricks(wall [][]int) int {
	if len(wall) == 0 {
		return 0
	}
	if len(wall[0]) == 0 {
		return 0
	}

	d := map[int]int{}

	for _, row := range wall {
		width := 0
		for i := 0; i < len(row); i++ {
			width += row[i]
			if i < len(row)-1 {
				if _, ok := d[width]; ok {
					d[width] += 1
				} else {
					d[width] = 1
				}
			}
		}
	}

	ans := len(wall)
	for _, val := range d {
		ans = Min(len(wall)-val, ans)
	}

	return ans
}
